### PR TITLE
[BEAM-12795] When using schemas, allow state/timers use without KV objects.

### PR DESCRIPTION
**Pull request by reuvenlax**
When using state/timers on schema PCollection, we want to avoid the need to convert first into a KV object. It's unnecessary, and the fact that KV doesn't have a schema (since KvCoder can't be a SchemaCoder) makes it awkward to use in a schema world.

We introduce a new StateKeyFields annotation that can be used inside a state/timers using DoFn to specify which fields of the input schema should be used as the key. For example, the following DoFn will keep state per user and country pair.

```java
pc.apply(new DoFn<InputT, OutputT>() {
  @StateKeyFields private final StateKeySpec keySpec = StateKeySpec.fields("userId", "location.country")
  final @StateId("state") StateSpec<ValueState<Integer>> stateSpec = StateSpecs.value();

  @ProcessElement
   public static process(@Element InputT input, @Key UserCountry userCountry) {
  }
});
```

Note that the `@Key` parameter annotation can be used to retrieve the current key. Any user type that has a schema matching the selected key can be used as a Key parameter (in the above example, the UserCountry type), and Beam will automatically convert the key to that type. If the key is a singleton field such as a String, you can also pass in the appropriate primitive type. You can also specify the parameter using a Row type; this all parallels the `@Element` parameter, which can also take any matching schema type.

One can also specify the key fields in the ParDo application, e.g. ParDo.of(doFn).withKeyFields("userId", "location.country"). This is useful when programmatically generating pipelines.

The automatic type conversion code already existed in DoFnSchemaInformation, and we simply extended that code to apply to key fields as well. We add a new primitive DoFn - ParDo.MultiOutputSchemaKeyFields - that is used only when key fields are specified. This is not a new primitive from the perspective of the portable model - it is simply a different type of ParDo, and shares the same URN.

### CODE REVIEW

**Comment by reuvenlax** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
friendly ping (though I realize there's a lot to unpack here)

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/EnvironmentsTest.java [hunk 1]
```diff
@@ -249,7 +238,7 @@ public void process(ProcessContext ctxt) {}
     Environment env1 = Environments.getEnvironment(ptransform, rehydratedComponents).get();
     assertThat(
         env1,
-        equalTo(components.toComponents().getEnvironmentsOrThrow(ptransform.getEnvironmentId())));
+        equalTo(components.toComponents().getEnvironmentsOrThrow(ptransform.getEnvironmentId())));*/
```
Comment (TheNeuralBit) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/EnvironmentsTest.java:241]: Is this intentional?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java [hunk 2]
```diff
@@ -100,7 +101,8 @@
               windowingStrategy,
               doFnSchemaInformation,
               sideInputMapping);
-      if (DoFnSignatures.signatureForDoFn(fn).usesState()) {
+      DoFnSignature signature = DoFnSignatures.signatureForDoFn(fn);
+      if (signature.usesState() || signature.onWindowExpiration() != null) {
```
Comment (TheNeuralBit) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java:105]: Is this relevant to this change?

**Review by TheNeuralBit — COMMENTED** [review]
I have a couple comments on odd things I noticed, but this would take me quite a while to review properly. The schema aspect is pretty straightforward, the integration with State/Timer APIs and core infrastructure (which I'm not very familiar with) is much more complex.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java (hunk 2 above)
Comment (reuvenlax) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java:105]: Yes, though it could also be separated. OnWindowExpiration didn't work if there were no state variables, which caused some of the tests I added to fail.

**Comment by lukecwik** [comment]
Finally got some time again to review this, taking a look now.

**Comment by lukecwik** [comment]
R: @lukecwik

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoSchemaTest.java [hunk 3]
```diff
@@ -681,30 +696,349 @@ public void processElement(
...
-                    KV.of("hello", Row.withSchema(type).addValue("b").build()),
-                    KV.of("hello", Row.withSchema(type).addValue("c").build()),
-                    KV.of("hello", Row.withSchema(type).addValue("d").build())))
-            .apply(ParDo.of(fn))
+                        Row.withSchema(type).addValues("a", "hello").build(),
+                        Row.withSchema(type).addValues("b", "hello").build(),
+                        Row.withSchema(type).addValues("c", "hello").build(),
+                        Row.withSchema(type).addValues("d", "hello").build())
+                    .withRowSchema(type))
+            .apply("run statetful fn", ParDo.of(fn).withOutputTags(mainTag, TupleTagList.empty()))
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoSchemaTest.java:710]:
```suggestion
            .apply("Run stateful fn", ParDo.of(fn).withOutputTags(mainTag, TupleTagList.empty()))
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java [hunk 4]
```diff
@@ -431,22 +436,37 @@ private static void finishSpecifyingStateSpecs(
...
     Coder<?> inputCoder = input.getCoder();
-    checkArgument(
-        inputCoder instanceof KvCoder,
-        "%s requires its input to use %s in order to use state and timers.",
-        ParDo.class.getSimpleName(),
-        KvCoder.class.getSimpleName());
+    if (fieldAccessDescriptor == null) {
+      checkArgument(
+          inputCoder instanceof KvCoder,
+          "%s requires its input to either use %s or have a schema input in order to use state and timers.",
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java:445]: Should we mention the `@StateKeyFields` annotation is necessary?

**Review by lukecwik — COMMENTED** [review]
This was quite difficult to review but I think this will work with pipeline update but I still strongly believe that this adds difficult to understand pipeline construction time complexity that we won't be able to remove. Even if we have a pipeline update story that allows users to export state they would need to be told on how to do a conversion of this kind which seems non-trivial.

I would still push for the implementation which relaxes the requirement on using a KV type on the GroupByKey in Dataflow and other runners which handle stateful DoFns to handle schematized types with stateful DoFn's that declare their keys.

Finally, it seems as though you had several minor (e.g. javadoc) and some seemingly important bug fixes that we should pull out of this PR and get submitted separately and much sooner.

**Comment by reuvenlax** [comment]
@lukecwik any more thoughts on this approach?
